doc/neps/ufunc-overrides.rst

```diff
@@ -153,49 +152,207 @@ tuple in the ``out`` keyword argument.
+``__array_ufunc__``, which implicitly define the type casting hierarchy,
+don't contradict this.
+
+It is useful to think of the typecasting hierarchy as a graph (see
+example below) in which, for any given class A, all other classes that
+define ``__array_ufunc__`` must belong to exactly one of three groups
+(making this an directed acyclic graph):
+
+- *Above A*: their ``__array_ufunc__`` can handle class A or some
+  member of the "above A" classes. In other words, these are the types
+  that A can be (indirectly) upcast to in ufuncs.
+
+- *Below A*: they can be handled by the ``__array_ufunc__`` of class A
+  or the ``__array_ufunc__`` of some member of the "below A" classes. In
+  other words, these are the types that can be (indirectly) upcast to A
+  in ufuncs.
+
+- *Incompatible*: neither above nor below A; types for which no
+  (indirect) upcasting is possible.
+
+Given this grouping, to ensure that expressions involving ufuncs either
+raise a :exc:`TypeError`, or have a result type that is independent of
+what ufuncs were called, what order they were called in, and what order
+their arguments were in, the above implies that ``__array_ufunc__`` for
+type A should:
+
+- Return an object of type A if all other arguments are of types below A.
+
+- Return :obj:`NotImplemented` if any argument has a type that is above
+  A or with which it is incompatible.
+
+Note that there are, as always, exceptions.  For instance, for a
+quantity class, the results of most ufuncs should be quantities, but
+this is not the case for comparison operators. For those, a quantity
+class would return a plain array.
+
+Note also that the legacy behaviour of numpy ufunc (legacy) behavior is
+to try to convert unknown objects to :class:`ndarray` via
+:func:`np.asarray`.  This is equivalent to placing :class:`ndarray` at
+the very top of the graph, and is thus a consistent type
+hierarchy (although one that causes the problems that motivate
+this NEP...).  By instead letting :class:`ndarray` return
+`NotImplemented` if any argument defines ``__array_ufunc__``, we provide
+the option for other classes to have :class:`ndarray` at the bottom of
+the type hierarchy.
+
+.. admonition:: Example
+
+   Type casting hierarchy
+
+   .. graphviz::
+
+      digraph array_ufuncs {
+         rankdir=BT;
+         A -> C;
+         B -> C;
+         D -> B;
+         ndarray -> A;
+         ndarray -> B;
+      }
+
+   The ``__array_ufunc__`` of type A can handle ndarrays, B can handle
+   ndarray and D, and C can handle A and B but not ndarrays or D.  The
+   result is a directed acyclic graph, and defines a type casting
+   hierarchy, with relations ``C > A > ndarray``, ``C > B > ndarray``,
+   ``C > B > D``. The type B is incompatible relative to A and vice
+   versa, and A and ndarray are incompatible relative to D.  Ufunc
+   expressions involving these classes should produce results of the
+   highest type involved or raise a :exc:`TypeError`.
```

Comment:
I added some additional examples of "bad" cases in my branch here that can be useful to add.
(The binop discussion there is not yet ready for discussion, that's just writing some thoughts down.).